### PR TITLE
Delete optimizer step in Fairseq's trainer

### PR DESCRIPTION
**Pull request by taylanbil**
Since we take a step in xla, this line must be removed.

Will need to update the examples repo with the correct commit of the fairseq dependency once this is merged.

### CODE REVIEW

File: fairseq/trainer.py [hunk 1]
```diff
@@ -344,7 +344,7 @@ def maybe_no_sync():
             self._prev_grad_norm = grad_norm
 
             # take an optimization step
-            self.optimizer.step()
+            # self.optimizer.step()  # XLA: this is done by torch_xla
```
Comment (dlibenzi) [inline fairseq/trainer.py]: Maybe adding a comment which explain where precisely this happens?
I see a function removed here, and nowhere close a matching `xm.optimizer_step()` call.

**Review by dlibenzi — CHANGES_REQUESTED** [review]

File: fairseq/trainer.py [hunk 2]
```diff
@@ -344,7 +345,12 @@ def maybe_no_sync():
             self._prev_grad_norm = grad_norm
 
             # take an optimization step
-            # self.optimizer.step()  # XLA: this is done by torch_xla
+
+            # xla takes care of optimization step using
+            #   torch_xla.xla_model.optimizer_step
+            # so skip optimization step here in that case
+            if not self.xla:
```
Comment (dlibenzi) [inline fairseq/trainer.py:352]: What if you let a lambda in, which defaults to optimizer.step(), which we override to an empty function?

Comment (taylanbil) [reply fairseq/trainer.py:352]: I can do that, but I thought I could use this flag for other types of behavior elsewhere in the object in the future. Would you like me to switch to lambda?

Comment (dlibenzi) [reply fairseq/trainer.py:352]: Up to you. I only have a limited view of the overall model.

**Review by dlibenzi — APPROVED** [review]

File: fairseq/trainer.py (hunk 2 above)
Comment (taylanbil) [reply fairseq/trainer.py:352]: ok let's go with this then, I can switch to lambda any time we want.

**Pull request merged**